### PR TITLE
Autogenerate api client

### PR DESCRIPTION
**Pull request by santaduda**
Autogenerate the typescript-axios api-client when the workflow is dispatched from tdei-gateway.

### CODE REVIEW

File: autogenerate_api_client.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
+curl -X POST https://generator3.swagger.io/api/generate -H 'content-type: application/json' -d '{ "specURL" : "https://tdei-gateway-dev.azurewebsites.net/v3/api-docs", "lang" : "typescript-axios", "type" : "CLIENT", "codegenVersion" : "V3"}' --output tdei-client-axios.zip
```
Comment (ayydevara) [inline autogenerate_api_client.sh:1]: is this endpoint `https://tdei-gateway-dev.azurewebsites.net/v3/api-docs` always available? 

can you make changes to this script to take the endpoint as input?

File: autogenerate_api_client.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  git checkout -b scratch/api-client
+  git add . 
+  git commit -m "system:auto generated API client"
+  git push -fu origin scratch/api-client
+else
+    echo "Git is not installed."
+fi
+
+if which gh >/dev/null 2>&1; then
+    gh pr create --base main --fill
```
Comment (ayydevara) [inline autogenerate_api_client.sh:15]: Are `gh` and `git` commands available on the pipelines?

File: autogenerate_api_client.sh (hunk 1 above)
Comment (susrisha) [inline autogenerate_api_client.sh:1]: This curl request does not specify the name of the repo. Needs to be modified. Eg.

```shell
curl --location 'https://generator3.swagger.io/api/generate' \
--header 'Content-Type: application/json' \
--data '{
    "specURL": "https://tdei-gateway-dev.azurewebsites.net/v3/api-docs",
    "lang": "typescript-axios",
    "type": "CLIENT",
    "codegenVersion": "V3",
    "options": {
        "additionalProperties": {
            "modelPropertyNaming": "original",
             "npmName":"tdei-client",
             "npmVersion":"1.0.0"
        }
    }
}' --output tdei-client.zip
```

Comment (susrisha) [reply autogenerate_api_client.sh:1]: Even in the above code, there should be a way to specify the npmVersion from user or auto increment

File: autogenerate_api_client.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+curl -X POST https://generator3.swagger.io/api/generate -H 'content-type: application/json' -d '{ "specURL" : "https://tdei-gateway-dev.azurewebsites.net/v3/api-docs", "lang" : "typescript-axios", "type" : "CLIENT", "codegenVersion" : "V3"}' --output tdei-client-axios.zip
+unzip -qo tdei-client-axios.zip
+
+if which git >/dev/null 2>&1; then
+  git branch -D scratch/api-client
```
Comment (susrisha) [inline autogenerate_api_client.sh:5]: why `scratch/api-client` branch?

**Review by susrisha — CHANGES_REQUESTED** [review]

File: autogenerate_api_client.sh (hunk 2 above)
Comment (santaduda) [reply autogenerate_api_client.sh:15]: It worked on my local machine so it should work in pipelines as well..

File: autogenerate_api_client.sh (hunk 1 above)
Comment (santaduda) [reply autogenerate_api_client.sh:1]: The api-client workflow is planned to be triggered automatically when the api gateway main/src contents is changed. Are we planing to make a change in the way the workflow is triggered?

Comment (santaduda) [reply autogenerate_api_client.sh:1]: Will see if I can read the npm version of the previous package.json and autoincrement it?

File: autogenerate_api_client.sh (hunk 3 above)
Comment (ayydevara) [reply autogenerate_api_client.sh:5]: @santaduda @susrisha 
Firstly, I used it as an example locally to verify the script was working.
Secondly, I mentioned in the call to either directly commit to the `main` branch or let someone review and merge it as PR although I prefer to do the latter as everything is automated. This is a choice we have to make as a team.

Finally, good point on bumping the npm version. We should do that!

File: autogenerate_api_client.sh (hunk 1 above)
Comment (ayydevara) [reply autogenerate_api_client.sh:1]: > This curl request does not specify the name of the repo. Needs to be modified. Eg.
> 
> ```shell
> curl --location 'https://generator3.swagger.io/api/generate' \
> --header 'Content-Type: application/json' \
> --data '{
>     "specURL": "https://tdei-gateway-dev.azurewebsites.net/v3/api-docs",
>     "lang": "typescript-axios",
>     "type": "CLIENT",
>     "codegenVersion": "V3",
>     "options": {
>         "additionalProperties": {
>             "modelPropertyNaming": "original",
>              "npmName":"tdei-client",
>              "npmVersion":"1.0.0"
>         }
>     }
> }' --output tdei-client.zip
> ```

You can replace it with this @santaduda. The one I gave you also works but specifying the additional properties while generating the client makes sense as we initially did the same

Comment (ayydevara) [reply autogenerate_api_client.sh:1]: @santaduda the question is about the gateway API endpoint, ideally, this is what should happen.

- A Commit/PR is merged to the gateway repo
- New Code is deployed to some environment
- GitHub Workflow validates if there are any changes to the `src/main` folder or maybe, we can further refine this only to validate if there are changes to any code files ` for e.g. (.java files changed)`
- If the above validation results in a success code, it dispatches an event to auto-generate a new API client
-  On the API Client repo, a listener listens to the above event and fires up the bash script for auto client generation

- The shell script should take two inputs
```sh 
bash autogenerate_api_client.sh <ENDPOINT> <npmVersion>
```

Comment (santaduda) [reply autogenerate_api_client.sh:1]: @ayydevara The above workflow is how it is implemented currently. 
I have updated the workflow to pass only 1 parameter <ENDPOINT> the <npmVersion> I can read it from the package.json and increment the version.

As per the github api documentation https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28 the workflow input can be passed with the "inputs" object.

**Comment by santaduda** [comment]
Updated the PR with new commit id [Receive end point from gateway api call](https://github.com/TaskarCenterAtUW/TDEI-api-client/pull/1/commits/d2f5704c45a357d7ca11d4111fb30c7e06a29279)

@ayydevara  @susrisha  Can you kindly review it ?

**Review by susrisha — APPROVED** [review]

**Pull request merged**